codegenerator/cli/templates/static/codegen/src/eventFetching/SourceManager.res

```diff
@@ -0,0 +1,176 @@
+open Belt
+
+let toQueryId = (query: FetchState.nextQuery) => {
+  query.fetchStateRegisterId->FetchState.registerIdToString ++ "-" ++ query.fromBlock->Int.toString
+}
+
+// Can't simply store fetching partitions, since fetchBatch
+// can be called with old chainFetchers after isFetching was set to false,
+// but state isn't still updated with fetched data.
+// This is temporary until all fetching logic moved to mutable state,
+// but for now prevent fetching the same partition twice with the same query,
+// using lastFetchedQueryId (aka idempotency key)
+type partitionFetchingState = {
+  isFetching: bool,
+  lastFetchedQueryId?: string,
+}
+
+// Ideally the ChainFetcher name suits this better
+// But currently the ChainFetcher module is immutable
+// and handles both processing and fetching.
+// So this module is to encapsulate the fetching logic only
+// with a mutable state for easier reasoning and testing.
+type t = {
+  logger: Pino.t,
+  maxPartitionConcurrency: int,
+  mutable isWaitingForNewBlock: bool,
+  mutable allPartitionsFetchingState: array<partitionFetchingState>,
+  // Don't use fetchingPartitions size, but have a separate counter
+  // to take into consideration partitions fetching for previous states (before rollback)
+  mutable fetchingPartitionsCount: int,
+  // Keep track on the current state id
+  // to work with correct state during rollbacks & preRegistration
+  mutable currentStateId: int,
+}
+
+let make = (~maxPartitionConcurrency, ~logger) => {
+  logger,
+  maxPartitionConcurrency,
+  isWaitingForNewBlock: false,
+  // Don't prefill with empty partitionFetchingState,
+  // since partitions might be added with the lifetime of the application.
+  // So lazily create fetchingState, when we execute a new partition
+  allPartitionsFetchingState: [],
+  currentStateId: 0,
+  fetchingPartitionsCount: 0,
+}
+
+exception FromBlockIsHigherThanToBlock({fromBlock: int, toBlock: int})
+
+let fetchBatch = async (
+  sourceManger: t,
+  ~allPartitions: PartitionedFetchState.allPartitions,
+  ~currentBlockHeight,
+  ~executePartitionQuery,
+  ~waitForNewBlock,
+  ~onNewBlock,
+  ~maxPerChainQueueSize,
+  ~setMergedPartitions,
+  ~stateId,
+) => {
```

Comment:
Even though it looks complicated, but now it has almost 100% test coverage, so I find it a drastic improvement.